{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ingest data from pipeline runs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We've now seen how individual datasets can be ingested, let's move on to ingesting datasets generated by a pipeline run. \n",
    "\n",
    "```{note}\n",
    "\n",
    "For the purpose of this guide, we ingest the pipeline output from within this notebook. Typically, this is done from the command line.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[`lnbfx`](https://lamin.ai/docs/lnbfx) is an open-source package to manage data from bioinformatics pipelines, complementary to workflow tools."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we show how to ingest a file from bfx runs, but is applicable to any types of pipeline runs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import lamindb as ln\n",
    "import lnbfx\n",
    "\n",
    "ln.nb.header()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we ingest a set of bioinformatics output files generated by [Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bfx_run_output = ln.datasets.dir_scrnaseq_cellranger()\n",
    "filepath = bfx_run_output / \"fastq/sample_1_R1.fastq.gz\"\n",
    "\n",
    "filepath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ingest = ln.db.ingest.add(filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create bfx pipeline entries to be inserted:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bfx_pipeline = lnbfx.lookup.pipeline.cell_ranger_v7_0_0\n",
    "pipeline = ln.db.query.pipeline(**bfx_pipeline).one_or_none()\n",
    "if pipeline is None:\n",
    "    pipeline_id = ln.db.insert.pipeline(**bfx_pipeline)\n",
    "    ln.db.insert.bfx_pipeline(id=bfx_pipeline[\"id\"], v=bfx_pipeline[\"v\"])\n",
    "else:\n",
    "    pipeline_id = pipeline.id\n",
    "\n",
    "# create a pipeline_run entry\n",
    "pipeline_run = ln.schema.core.pipeline_run(\n",
    "    pipeline_id=pipeline_id, pipeline_v=bfx_pipeline[\"v\"], name=\"bfx_run_001\"\n",
    ")\n",
    "\n",
    "# create a bfx_run entry\n",
    "bfx_run = ln.schema.bfx.bfx_run(\n",
    "    id=pipeline_run.id,\n",
    "    dir=\"bfx_run_001\",\n",
    "    bfx_pipeline_id=bfx_pipeline[\"id\"],\n",
    "    bfx_pipeline_v=bfx_pipeline[\"v\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pipeline_run, bfx_run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Link the pipeline entries to dobject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ingest.link.pipeline_run(pipeline_run)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ingest.link.add_entry(\"bfx_run\", bfx_run)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Link biometa to dobject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_id = ln.db.insert.biosample(name=\"test_biosample\")\n",
    "biometa = ln.schema.wetlab.biometa(biosample_id=biosample_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "biometa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ingest.link.biometa(biometa)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check all linked entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ingest.link.linked_entries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Complete the ingestion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ingest.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Query dobject by linked metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ln.db.query.dobject(where=dict(biosample=dict(name=\"test_biosample\"))).df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ln.db.query.dobject(where=dict(pipeline_run=dict(name=\"bfx_run_001\"))).df()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "nbproject": {
   "id": "61IJ1lyUDTQG",
   "parent": null,
   "pypackage": null,
   "time_init": "2022-10-07T23:07:51.916670+00:00",
   "version": "draft"
  },
  "vscode": {
   "interpreter": {
    "hash": "ae1fefc8646a06dd2e75004cd934adda7c5727b046986a772e3b44b0ffba9754"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
