{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50466c93",
   "metadata": {},
   "source": [
    "# Conceive experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7e482c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import lamindb as ln\n",
    "\n",
    "ln.nb.header()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "655e948d",
   "metadata": {},
   "source": [
    "This notebook is generated by `testuser1`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c551eb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "!lndb login testuser1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4ba56e6",
   "metadata": {},
   "source": [
    "## Register experiment level metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c97c60a0",
   "metadata": {},
   "source": [
    "Let's select for our project of interest:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1329b60",
   "metadata": {},
   "outputs": [],
   "source": [
    "project = ln.db.select.project(external_id=\"P003\").one()\n",
    "project"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c8b0836",
   "metadata": {},
   "source": [
    "Next, we register two experiment types:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb93306f",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_type1 = ln.db.insert.experiment_type(efo_id=\"EFO:0030033\", name=\"gRNA-seq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fb33754",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_type2 = ln.db.insert.experiment_type(efo_id=\"EFO:0008860\", name=\"Perturb-Seq\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90650cee",
   "metadata": {},
   "source": [
    "Next, we register two experiments that will be performed in this project:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "049d8000",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment1 = ln.db.insert.experiment(\n",
    "    external_id=\"EXP001\",\n",
    "    name=(\n",
    "        \"Genome-wide CRISPRi screen for functional regulators of cytokine production in\"\n",
    "        \" response to stimulation\"\n",
    "    ),\n",
    "    project_id=project.id,\n",
    "    experiment_type_id=exp_type1.id,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f364dd1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment2 = ln.db.insert.experiment(\n",
    "    external_id=\"EXP002\",\n",
    "    name=(\n",
    "        \"CRISPRa Perturb-seq for characterizing molecular phenotypes of cytokine\"\n",
    "        \" regulators\"\n",
    "    ),\n",
    "    project_id=project.id,\n",
    "    experiment_type_id=exp_type2.id,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2760be19",
   "metadata": {},
   "source": [
    "## Register biosample level metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "620c0a9f",
   "metadata": {},
   "source": [
    "Next, we register a biosample that will be used in the experiments:\n",
    "\n",
    "Along with some biological entities that will be used for annotating the biosample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "245b1f1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = ln.db.insert.species(common_name=\"human\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f3e3c27",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = ln.db.insert.cell_type(ontology_id=\"CL:0000084\", name=\"T cell\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba10bd72",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue = ln.db.insert.tissue(\n",
    "    ontology_id=\"CL:2000001\", name=\"peripheral blood mononuclear cell\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a27dd22",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample = ln.db.insert.biosample(\n",
    "    external_id=\"S1\",\n",
    "    name=\"primary human T cell\",\n",
    "    batch=\"1\",\n",
    "    species_id=species.id,\n",
    "    cell_type_id=cell_type.id,\n",
    "    tissue_id=tissue.id,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5be95da",
   "metadata": {},
   "source": [
    "## Register readouts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77c1d1b5",
   "metadata": {},
   "source": [
    "Next, we register two types of readout that will be captured:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75535150",
   "metadata": {},
   "outputs": [],
   "source": [
    "readout1 = ln.db.insert.readout(efo_id=\"EFO:0003024\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84c0b062",
   "metadata": {},
   "outputs": [],
   "source": [
    "readout2 = ln.db.insert.readout(efo_id=\"EFO:0008913\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47e1d597",
   "metadata": {},
   "source": [
    "## Register featuresets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "202a5144",
   "metadata": {},
   "source": [
    "Here we also register two sets of features that will be measured in the experiments:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ec2eadf",
   "metadata": {},
   "outputs": [],
   "source": [
    "featureset1 = ln.db.insert.featureset(feature_entity=\"gene\", name=\"CRISPRa-geneset-1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1472a739",
   "metadata": {},
   "outputs": [],
   "source": [
    "featureset2 = ln.db.insert.featureset(\n",
    "    feature_entity=\"gene\", name=\"Perturbseq-geneset-1\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eefd6e0",
   "metadata": {},
   "source": [
    "## Register biometa\n",
    "\n",
    "biometa holds the following 4 fields and is directly link to datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0415107",
   "metadata": {},
   "outputs": [],
   "source": [
    "biometa1_id = ln.db.insert.biometa(\n",
    "    experiment_id=experiment1.id,\n",
    "    biosample_id=biosample.id,\n",
    "    readout_id=readout1.id,\n",
    "    featureset_id=featureset1.id,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "692c16e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "biometa2_id = ln.db.insert.biometa(\n",
    "    experiment_id=experiment2.id,\n",
    "    biosample_id=biosample.id,\n",
    "    readout_id=readout2.id,\n",
    "    featureset_id=featureset2.id,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "nbproject": {
   "id": "W4oWHHbY0oG6",
   "parent": null,
   "pypackage": null,
   "time_init": "2022-08-30T22:19:33.279387+00:00",
   "version": "draft"
  },
  "vscode": {
   "interpreter": {
    "hash": "ae1fefc8646a06dd2e75004cd934adda7c5727b046986a772e3b44b0ffba9754"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
